Reorganize SoftDevices for NRF52 series

marcuschangarm opened the pull request:
### Description

New directory structure:

 * TARGET_SOFTDEVICE_COMMON
 * TARGET_SOFTDEVICE_S112
 * TARGET_SOFTDEVICE_S132_FULL (MBR + SoftDevice, default)
 * TARGET_SOFTDEVICE_S132_OTA (SoftDevice only, for firmware updates)
 * TARGET_SOFTDEVICE_S132_MBR (MBR only, for bootloader builds)
 * TARGET_SOFTDEVICE_S140_FULL (MBR + SoftDevice, default)
 * TARGET_SOFTDEVICE_S140_OTA (SoftDevice only, for firmware updates)
 * TARGET_SOFTDEVICE_S140_MBR (MBR only, for bootloader builds)
 * TARGET_SOFTDEVICE_NONE

The X_OTA and X_MBR binaries are obtained from the original x_FULL SoftDevice
by splitting it in an MBR part and a SoftDevice part. The MBR is needed for
the bootloader and the SoftDevice for firmware updates.

Build application without SoftDevice:

```
  "target_overrides": {
      "*": {
          "target.extra_labels_remove": ["SOFTDEVICE_COMMON", "SOFTDEVICE_X_FULL"],
          "target.extra_labels_add": ["SOFTDEVICE_NONE"]
      }
  }
```

Build application for firmware update using SoftDevice X:

```
  "target_overrides": {
      "*": {
          "target.extra_labels_remove": ["SOFTDEVICE_X_FULL"],
          "target.extra_labels_add": ["SOFTDEVICE_X_OTA"]
      }
  }
```

Build bootloader without SoftDevice X:

```
  "target_overrides": {
      "*": {
          "target.extra_labels_remove": ["SOFTDEVICE_X_FULL"],
          "target.extra_labels_add": ["SOFTDEVICE_X_MBR"]
      }
  }
```

<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/latest/reference/workflow.html (Pull request template)
-->


### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [x] Refactor
    [ ] New target
    [ ] Feature
    [x] Breaking change

0xc0170 commented:
@marcuschangarm How this breaks the current apps? Can it be made backward compatible ?

marcuschangarm commented:
Its not more breaking than the 5.9 stuff we've been doing! 😄
Just wanted to make sure it got tagged correctly.

cmonr commented:
There sure are a bunch of new hex files and bootloaders. Would it be possible to comment in the commit message as to how they were generated?

cmonr commented:
> Build application without SoftDevice:

Shouldn't this be the default action for Nordic devices instead of needing to explicity add it to target_overrides?

marcuschangarm commented:
Added: 
```
The x_OTA and x_MBR binaries are obtained from the original x_FULL SoftDevice
by splitting it in an MBR part and a SoftDevice part. The MBR is needed for
the bootloader and the SoftDevice for firmware updates.
```

marcuschangarm commented:
>>Build application without SoftDevice:

>Shouldn't this be the default action for Nordic devices instead of needing to explicity add it to target_overrides?

That would be a deviation from what we have done since the first NRF51.

cmonr commented:
/morph build

cmonr commented:
/morph build

studavekar commented:
/morph build

cmonr commented:
Looks like something different went wrong with the CI.

marcuschangarm commented:
/morph build

marcuschangarm commented:
/morph build

marcuschangarm commented:
/morph test

marcuschangarm commented:
/morph build

cmonr commented:
@donatieng When you get a chance.

Most of the changes are either moves or binary diffs.

marcuschangarm commented:
/morph build

marcuschangarm commented:
/morph build

marcuschangarm commented:
/morph build

cmonr commented:
Restarting build since statuses appear to not have reported properly.

/morph build

cmonr commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_SDK_14_2/TARGET_SD_S132/mbed_lib.json:
```diff
@@ -0,0 +1,43 @@
+    "name": "softdevice",
+    "target_overrides": {
+        "*": {
+            "target.macros_add": [
```
I find it hard to believe that all of these macros are needed.

cmonr commented on targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_SDK_14_2/TARGET_SOFTDEVICE_S132/mbed_lib.json:
```diff
@@ -1,48 +0,0 @@
-    "name": "softdevice",
-    "config": {
-    },    
-    "macros": [
```
Ah, and now I see why they were needed in the first place.

cmonr reviewed (dismissed):
Summary of changes:
- Most are files being moved.
- New hex and license files added.
- Softtarget-specific json files added.
- Updated macros.

Commit:
Trim white space and clean up NRF52 configuration files

Removed unused and duplicate macros.

Commit:
Add NRF52 support for building bootloader and updateable firmware

New directory structure:

 * TARGET_SOFTDEVICE_COMMON
 * TARGET_SOFTDEVICE_S112
 * TARGET_SOFTDEVICE_S132_FULL (MBR + SoftDevice, default)
 * TARGET_SOFTDEVICE_S132_OTA (SoftDevice only, for firmware updates)
 * TARGET_SOFTDEVICE_S132_MBR (MBR only, for bootloader builds)
 * TARGET_SOFTDEVICE_S140_FULL (MBR + SoftDevice, default)
 * TARGET_SOFTDEVICE_S140_OTA (SoftDevice only, for firmware updates)
 * TARGET_SOFTDEVICE_S140_MBR (MBR only, for bootloader builds)
 * TARGET_SOFTDEVICE_NONE

The X_OTA and X_MBR binaries are obtained from the original x_FULL SoftDevice
by splitting it in an MBR part and a SoftDevice part. The MBR is needed for
the bootloader and the SoftDevice for firmware updates.

Build application without SoftDevice:

  "target_overrides": {
      "*": {
          "target.extra_labels_remove": ["SOFTDEVICE_COMMON", "SOFTDEVICE_X_FULL"],
          "target.extra_labels_add": ["SOFTDEVICE_NONE"]
      }
  }

Build application for firmware update using SoftDevice X:

  "target_overrides": {
      "*": {
          "target.extra_labels_remove": ["SOFTDEVICE_X_FULL"],
          "target.extra_labels_add": ["SOFTDEVICE_X_OTA"]
      }
  }

Build bootloader without SoftDevice X:

  "target_overrides": {
      "*": {
          "target.extra_labels_remove": ["SOFTDEVICE_COMMON", "SOFTDEVICE_X_FULL"],
          "target.extra_labels_add": ["SOFTDEVICE_X_MBR"]
      }
  }

0xc0170 reviewed (dismissed):
Note, with all these updates, we need to make sure the release notes contain sufficient details for these changes (how nrf52 update from the current minor release to the one that this is in - as there are quite changes and won't be that trivial, will it?).

donatieng reviewed (commented):
Looks good - @marcuschangarm any chance you could add a README explaining how you split up the MBR/Softdevice?

0xc0170 commented:
The failure for export is the same one as previous one. We have seen it on master, was fixed a week ago. Can you test locally on your branch, or shall it be rebased (however CI merges this with the destination branch) so should not be there at all.

@theotherjimmy @studavekar Please look at this

studavekar commented:
> The failure for export is the same one as previous one. We have seen it on master, was fixed a week ago. Can you test locally on your branch, or shall it be rebased (however CI merges this with the destination branch) so should not be there at all.
> 
> @theotherjimmy @studavekar Please look at this

The fix https://github.com/ARMmbed/mbed-os/pull/6713 for error `startup_nrf52832.S: error: A3907U: Via file '.\build\startup_nrf52832._ia' command too long for buffer`. 

looks like still the commandline is too long to handle.

marcuschangarm commented:
I'm undoing the nested folder structure and keeping the directories flat. The PR will be about adding the MBR and OTA binaries.

marcuschangarm commented:
@donatieng 

> any chance you could add a README explaining how you split up the MBR/Softdevice?

Sure, I literally just copy pasted it in an intel hex editor.

marcuschangarm commented:
@0xc0170 

> Note, with all these updates, we need to make sure the release notes contain sufficient details for these changes (how nrf52 update from the current minor release to the one that this is in - as there are quite changes and won't be that trivial, will it?).

I can collect a list with changes. The major changes have been to things that didn't work correctly to begin or to people that for some reason have linked directly to files in the SDK.

marcuschangarm commented:
/morph build

marcuschangarm commented:
/morph test
/morph export-build

marcuschangarm commented:
/morph test

marcuschangarm commented:
/morph test

marcuschangarm commented:
/morph build

cmonr commented:
/morph test
/morph export-build
